Bottom: 35f384b602b56188382fb57fc7834ce82df0edb8
Top:    10671b8c1ff7c82923b4f956b67ea210d153bed8
Author: Ben Greear <greearb@candelatech.com>
Date:   2013-05-09 11:56:24 -0700

netdev: Add pktgen hooks to dev.c rx logic.

Signed-off-by: Ben Greear <greearb@candelatech.com>


---

diff --git a/net/core/dev.c b/net/core/dev.c
index 978ac8f..7971dd3 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -148,6 +148,23 @@
 /* This should be increased if a protocol with a bigger head is added. */
 #define GRO_MAX_HEAD (MAX_HEADER + 128)
 
+#if defined(CONFIG_NET_PKTGEN) || defined(CONFIG_NET_PKTGEN_MODULE)
+#include "pktgen.h"
+
+#warning "Compiling dev.c for pktgen.";
+
+int (*handle_pktgen_hook)(struct sk_buff *skb) = NULL;
+EXPORT_SYMBOL(handle_pktgen_hook);
+
+static __inline__ int handle_pktgen_rcv(struct sk_buff* skb) {
+	if (handle_pktgen_hook) {
+		return handle_pktgen_hook(skb);
+	}
+	return -1;
+}
+#endif
+
+
 static DEFINE_SPINLOCK(ptype_lock);
 static DEFINE_SPINLOCK(offload_lock);
 struct list_head ptype_base[PTYPE_HASH_SIZE] __read_mostly;
@@ -4194,6 +4211,17 @@ ncls:
 		skb->vlan_tci = 0;
 	}
 
+#if defined(CONFIG_NET_PKTGEN) || defined(CONFIG_NET_PKTGEN_MODULE)
+	if ((skb->dev->pkt_dev) &&
+	    (handle_pktgen_rcv(skb) >= 0)) {
+		/* Pktgen may consume the packet, no need to send
+		 * to further protocols.
+		 */
+		ret = NET_RX_SUCCESS;
+		goto out;
+	}
+#endif
+
 	type = skb->protocol;
 
 	/* deliver only exact match when indicated */
